File: Sources/XMTP/ConversationV2.swift
//
//  ConversationV2.swift
//
//
//  Created by Pat Nakajima on 11/26/22.
//

import CryptoKit
import Foundation

// Save the non-client parts for a v2 conversation
public struct ConversationV2Container: Codable {
	var topic: String
	var keyMaterial: Data
	var conversationID: String?
	var metadata: [String: String] = [:]
	var peerAddress: String
	var header: SealedInvitationHeaderV1

	public func decode(with client: Client) -> ConversationV2 {
		let context = InvitationV1.Context(conversationID: conversationID ?? "", metadata: metadata)
		return ConversationV2(topic: topic, keyMaterial: keyMaterial, context: context, peerAddress: peerAddress, client: client, header: header)
	}
}

/// Handles V2 Message conversations.
public struct ConversationV2 {
	public var topic: String
	public var keyMaterial: Data // MUST be kept secret
	public var context: InvitationV1.Context
	public var peerAddress: String
	public var client: Client
	private var header: SealedInvitationHeaderV1

	static func create(client: Client, invitation: InvitationV1, header: SealedInvitationHeaderV1) throws -> ConversationV2 {
		let myKeys = client.keys.getPublicKeyBundle()

		let peer = try myKeys.walletAddress == (try header.sender.walletAddress) ? header.recipient : header.sender
		let peerAddress = try peer.walletAddress

		let keyMaterial = Data(invitation.aes256GcmHkdfSha256.keyMaterial)

		return ConversationV2(
			topic: invitation.topic,
			keyMaterial: keyMaterial,
			context: invitation.context,
			peerAddress: peerAddress,
			client: client,
			header: header
		)
	}

	public init(topic: String, keyMaterial: Data, context: InvitationV1.Context, peerAddress: String, client: Client) {
		self.topic = topic
		self.keyMaterial = keyMaterial
		self.context = context
		self.peerAddress = peerAddress
		self.client = client
		header = SealedInvitationHeaderV1()
	}

	public init(topic: String, keyMaterial: Data, context: InvitationV1.Context, peerAddress: String, client: Client, header: SealedInvitationHeaderV1) {
		self.topic = topic
		self.keyMaterial = keyMaterial
		self.context = context
		self.peerAddress = peerAddress
		self.client = client
		self.header = header
	}

	public var encodedContainer: ConversationV2Container {
		ConversationV2Container(topic: topic, keyMaterial: keyMaterial, conversationID: context.conversationID, metadata: context.metadata, peerAddress: peerAddress, header: header)
	}

	func prepareMessage(encodedContent: EncodedContent, options: SendOptions?) async throws -> PreparedMessage {
		let message = try await MessageV2.encode(
			client: client,
			content: encodedContent,
			topic: topic,
			keyMaterial: keyMaterial
		)

		let topic = options?.ephemeral == true ? ephemeralTopic : topic

		let envelope = Envelope(topic: topic, timestamp: Date(), message: try Message(v2: message).serializedData())
		return PreparedMessage(envelopes: [envelope])
	}

	func prepareMessage<T>(content: T, options: SendOptions?) async throws -> PreparedMessage {
		let codec = client.codecRegistry.find(for: options?.contentType)

		func encode<Codec: ContentCodec>(codec: Codec, content: Any) throws -> EncodedContent {
			if let content = content as? Codec.T {
				return try codec.encode(content: content, client: client)
			} else {
				throw CodecError.invalidContent
			}
		}

		var encoded = try encode(codec: codec, content: content)
        
        func fallback<Codec: ContentCodec>(codec: Codec, content: Any) throws -> String? {
            if let content = content as? Codec.T {
                return try codec.fallback(content: content)
            } else {
                throw CodecError.invalidContent
            }
        }
        
        if let fallback = try fallback(codec: codec, content: content) {
            encoded.fallback = fallback
        }
		
        if let compression = options?.compression {
			encoded = try encoded.compress(compression)
		}

		return try await prepareMessage(encodedContent: encoded, options: options)
	}

	func messages(limit: Int? = nil, before: Date? = nil, after: Date? = nil, direction: PagingInfoSortDirection? = .descending) async throws -> [DecodedMessage] {
		let pagination = Pagination(limit: limit, before: before, after: after, direction: direction)
		let envelopes = try await client.apiClient.envelopes(topic: topic.description, pagination: pagination)

		return envelopes.compactMap { envelope in
			do {
				return try decode(envelope: envelope)
			} catch {
				print("Error decoding envelope \(error)")
				return nil
			}
		}
	}

	func decryptedMessages(limit: Int? = nil, before: Date? = nil, after: Date? = nil, direction: PagingInfoSortDirection? = .descending) async throws -> [DecryptedMessage] {
		let pagination = Pagination(limit: limit, before: before, after: after, direction: direction)
		let envelopes = try await client.apiClient.envelopes(topic: topic.description, pagination: pagination)

		return try envelopes.map { envelope in
			let message = try Message(serializedData: envelope.message)
			return try MessageV2.decrypt(generateID(from: envelope), topic, message.v2, keyMaterial: keyMaterial, client: client)
		}
	}

	var ephemeralTopic: String {
		topic.replacingOccurrences(of: "/xmtp/0/m", with: "/xmtp/0/mE")
	}

	public func streamEphemeral() -> AsyncThrowingStream<Envelope, Error> {
		AsyncThrowingStream { continuation in
			Task {
				do {
					for try await envelope in client.subscribe(topics: [ephemeralTopic]) {
						continuation.yield(envelope)
					}
				} catch {
					continuation.finish(throwing: error)
				}
			}
		}
	}

	public func streamMessages() -> AsyncThrowingStream<DecodedMessage, Error> {
		AsyncThrowingStream { continuation in
			Task {
				for try await envelope in client.subscribe(topics: [topic.description]) {
					let decoded = try decode(envelope: envelope)

					continuation.yield(decoded)
				}
			}
		}
	}

	public var createdAt: Date {
		Date(timeIntervalSince1970: Double(header.createdNs / 1_000_000) / 1000)
	}

	public func decode(envelope: Envelope) throws -> DecodedMessage {
		let message = try Message(serializedData: envelope.message)

		return try MessageV2.decode(generateID(from: envelope), topic, message.v2, keyMaterial: keyMaterial, client: client)
	}

	@discardableResult func send<T>(content: T, options: SendOptions? = nil) async throws -> String {
		let preparedMessage = try await prepareMessage(content: content, options: options)
        return try await send(prepared: preparedMessage)
	}

	@discardableResult func send(content: String, options: SendOptions? = nil, sentAt _: Date) async throws -> String {
		let preparedMessage = try await prepareMessage(content: content, options: options)
        return try await send(prepared: preparedMessage)
	}

	@discardableResult func send(encodedContent: EncodedContent, options: SendOptions? = nil) async throws -> String {
		let preparedMessage = try await prepareMessage(encodedContent: encodedContent, options: options)
        return try await send(prepared: preparedMessage)
	}

    @discardableResult func send(prepared: PreparedMessage) async throws -> String {
        try await client.publish(envelopes: prepared.envelopes)
        if((await client.contacts.consentList.state(address: peerAddress)) == .unknown) {
            try await client.contacts.allow(addresses: [peerAddress])
        }
        return prepared.messageID
    }

	public func encode<Codec: ContentCodec, T>(codec: Codec, content: T) async throws -> Data where Codec.T == T {
		let content = try codec.encode(content: content, client: client)

		let message = try await MessageV2.encode(
			client: client,
			content: content,
			topic: topic,
			keyMaterial: keyMaterial
		)

		let envelope = Envelope(
			topic: topic,
			timestamp: Date(),
			message: try Message(v2: message).serializedData()
		)

		return try envelope.serializedData()
	}

	@discardableResult func send(content: String, options: SendOptions? = nil) async throws -> String {
		return try await send(content: content, options: options, sentAt: Date())
	}

	private func generateID(from envelope: Envelope) -> String {
		Data(SHA256.hash(data: envelope.message)).toHex
	}
}

File: library/src/main/java/org/xmtp/android/library/ConversationV2.kt
package org.xmtp.android.library

import android.util.Log
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.mapNotNull
import kotlinx.coroutines.runBlocking
import org.web3j.crypto.Hash
import org.xmtp.android.library.codecs.ContentCodec
import org.xmtp.android.library.codecs.EncodedContent
import org.xmtp.android.library.codecs.compress
import org.xmtp.android.library.messages.DecryptedMessage
import org.xmtp.android.library.messages.Envelope
import org.xmtp.android.library.messages.EnvelopeBuilder
import org.xmtp.android.library.messages.Message
import org.xmtp.android.library.messages.MessageBuilder
import org.xmtp.android.library.messages.MessageV2Builder
import org.xmtp.android.library.messages.Pagination
import org.xmtp.android.library.messages.PagingInfoSortDirection
import org.xmtp.android.library.messages.SealedInvitationHeaderV1
import org.xmtp.android.library.messages.getPublicKeyBundle
import org.xmtp.android.library.messages.walletAddress
import org.xmtp.proto.message.api.v1.MessageApiOuterClass
import org.xmtp.proto.message.contents.Invitation
import java.util.Date

data class ConversationV2(
    val topic: String,
    val keyMaterial: ByteArray,
    val context: Invitation.InvitationV1.Context,
    val peerAddress: String,
    val client: Client,
    val createdAtNs: Long? = null,
    private val header: SealedInvitationHeaderV1,
) {

    companion object {
        fun create(
            client: Client,
            invitation: Invitation.InvitationV1,
            header: SealedInvitationHeaderV1,
        ): ConversationV2 {
            val myKeys = client.keys.getPublicKeyBundle()
            val peer =
                if (myKeys.walletAddress == (header.sender.walletAddress)) header.recipient else header.sender
            val peerAddress = peer.walletAddress
            val keyMaterial = invitation.aes256GcmHkdfSha256.keyMaterial.toByteArray()
            return ConversationV2(
                topic = invitation.topic,
                keyMaterial = keyMaterial,
                context = invitation.context,
                peerAddress = peerAddress,
                client = client,
                createdAtNs = header.createdNs,
                header = header,
            )
        }
    }

    val createdAt: Date = Date((createdAtNs ?: header.createdNs) / 1_000_000)

    /**
     * This lists messages sent to the [Conversation].
     * @param before initial date to filter
     * @param after final date to create a range of dates and filter
     * @param limit is the number of result that will be returned
     * @param direction is the way of srting the information, by default is descending, you can
     * know more about it in class [MessageApiOuterClass].
     * @see MessageApiOuterClass.SortDirection
     * @return The list of messages sent. If [before] or [after] are specified then this will only list messages
     * sent at or [after] and at or [before].
     * If [limit] is specified then results are pulled in pages of that size.
     * If [direction] is specified then that will control the sort order of te messages.
     * @see Conversation.messages
     */
    fun messages(
        limit: Int? = null,
        before: Date? = null,
        after: Date? = null,
        direction: PagingInfoSortDirection = MessageApiOuterClass.SortDirection.SORT_DIRECTION_DESCENDING,
    ): List<DecodedMessage> {
        val pagination =
            Pagination(limit = limit, before = before, after = after, direction = direction)
        val result = runBlocking {
            client.apiClient.envelopes(
                topic = topic,
                pagination = pagination,
            )
        }

        return result.mapNotNull { envelope ->
            decodeEnvelopeOrNull(envelope)
        }
    }

    /**
     * This lists decrypted messages sent to the [Conversation].
     * @param before initial date to filter
     * @param after final date to create a range of dates and filter
     * @param limit is the number of result that will be returned
     * @param direction is the way of srting the information, by default is descending, you can
     * know more about it in class [MessageApiOuterClass].
     * @see MessageApiOuterClass.SortDirection
     * @return The list of messages sent. If [before] or [after] are specified then this will only list messages
     * sent at or [after] and at or [before].
     * If [limit] is specified then results are pulled in pages of that size.
     * If [direction] is specified then that will control the sort order of te messages.
     */
    fun decryptedMessages(
        limit: Int? = null,
        before: Date? = null,
        after: Date? = null,
        direction: PagingInfoSortDirection = MessageApiOuterClass.SortDirection.SORT_DIRECTION_DESCENDING,
    ): List<DecryptedMessage> {
        val pagination =
            Pagination(limit = limit, before = before, after = after, direction = direction)
        val envelopes = runBlocking { client.apiClient.envelopes(topic, pagination) }

        return envelopes.map { envelope ->
            decrypt(envelope)
        }
    }

    /**
     * This decrypts a message
     * @param envelope Object that contains all the information of the encrypted message
     * @return [DecryptedMessage] object
     */
    fun decrypt(envelope: Envelope): DecryptedMessage {
        val message = Message.parseFrom(envelope.message)
        return MessageV2Builder.buildDecrypt(
            id = generateId(envelope = envelope),
            topic,
            message.v2,
            keyMaterial,
            client,
        )
    }

    fun streamMessages(): Flow<DecodedMessage> = flow {
        client.subscribe(listOf(topic)).mapNotNull { decodeEnvelopeOrNull(envelope = it) }.collect {
            emit(it)
        }
    }

    /**
     * This encrypts a message
     * @param envelope Object that contains all the information of the decrypted message
     * @return [DecodedMessage] object
     */
    fun decodeEnvelope(envelope: Envelope): DecodedMessage {
        val message = Message.parseFrom(envelope.message)
        return MessageV2Builder.buildDecode(
            generateId(envelope = envelope),
            topic = topic,
            message.v2,
            keyMaterial = keyMaterial,
            client = client,
        )
    }

    /**
     * This encrypts a message
     * @param envelope Object that contains all the information of the decrypted message
     * @return [DecodedMessage] object if is not possible will return null
     */
    private fun decodeEnvelopeOrNull(envelope: Envelope): DecodedMessage? {
        return try {
            decodeEnvelope(envelope)
        } catch (e: Exception) {
            Log.d("CONV_V2", "discarding message that failed to decode", e)
            null
        }
    }

    suspend fun <T> send(content: T, options: SendOptions? = null): String {
        val preparedMessage = prepareMessage(content = content, options = options)
        return send(preparedMessage)
    }

    suspend fun send(text: String, options: SendOptions? = null, sentAt: Date? = null): String {
        val preparedMessage = prepareMessage(content = text, options = options)
        return send(preparedMessage)
    }

    suspend fun send(encodedContent: EncodedContent, options: SendOptions?): String {
        val preparedMessage = prepareMessage(encodedContent = encodedContent, options = options)
        return send(preparedMessage)
    }

    suspend fun send(prepared: PreparedMessage): String {
        client.publish(envelopes = prepared.envelopes)
        if (client.contacts.consentList.state(address = peerAddress) == ConsentState.UNKNOWN) {
            client.contacts.allow(addresses = listOf(peerAddress))
        }
        return prepared.messageId
    }

    fun <Codec : ContentCodec<T>, T> encode(codec: Codec, content: T): ByteArray {
        val encodedContent = codec.encode(content = content)
        val message = MessageV2Builder.buildEncode(
            client = client,
            encodedContent = encodedContent,
            topic = topic,
            keyMaterial = keyMaterial,
            codec = codec,
        )
        val envelope = EnvelopeBuilder.buildFromString(
            topic = topic,
            timestamp = Date(),
            message = MessageBuilder.buildFromMessageV2(v2 = message.messageV2).toByteArray(),
        )
        return envelope.toByteArray()
    }

    fun <T> prepareMessage(content: T, options: SendOptions?): PreparedMessage {
        val codec = Client.codecRegistry.find(options?.contentType)

        fun <Codec : ContentCodec<T>> encode(codec: Codec, content: Any?): EncodedContent {
            val contentType = content as? T
            if (contentType != null) {
                return codec.encode(contentType)
            } else {
                throw XMTPException("Codec type is not registered")
            }
        }

        var encoded = encode(codec = codec as ContentCodec<T>, content = content)
        val fallback = codec.fallback(content)
        if (!fallback.isNullOrBlank()) {
            encoded = encoded.toBuilder().also {
                it.fallback = fallback
            }.build()
        }
        val compression = options?.compression
        if (compression != null) {
            encoded = encoded.compress(compression)
        }
        return prepareMessage(encoded, options = options)
    }

    fun prepareMessage(
        encodedContent: EncodedContent,
        options: SendOptions?,
    ): PreparedMessage {
        val codec = Client.codecRegistry.find(options?.contentType)
        val message = MessageV2Builder.buildEncode(
            client = client,
            encodedContent = encodedContent,
            topic = topic,
            keyMaterial = keyMaterial,
            codec = codec,
        )

        val newTopic = if (options?.ephemeral == true) ephemeralTopic else topic

        val envelope = EnvelopeBuilder.buildFromString(
            topic = newTopic,
            timestamp = Date(),
            message = MessageBuilder.buildFromMessageV2(v2 = message.messageV2).toByteArray(),
        )
        return PreparedMessage(listOf(envelope))
    }

    private fun generateId(envelope: Envelope): String =
        Hash.sha256(envelope.message.toByteArray()).toHex()

    val ephemeralTopic: String
        get() = topic.replace("/xmtp/0/m", "/xmtp/0/mE")

    fun streamEphemeral(): Flow<Envelope> = flow {
        client.subscribe(listOf(ephemeralTopic)).collect {
            emit(it)
        }
    }

    fun streamDecryptedMessages(): Flow<DecryptedMessage> = flow {
        client.subscribe(listOf(topic)).collect {
            emit(decrypt(envelope = it))
        }
    }
}
